### PR TITLE
docs: add integration testing ADR

### PR DESCRIPTION
**Pull request by Henrrypg**
## **Summary**

  This ADR proposes a live LLM provider integration test suite for openedx-ai-extensions. The existing test suite mocks every LLM call, which means it cannot
  catch provider-side breakage: deprecated model names, schema enforcement changes, thread expiry, streaming incompatibilities, or semantically wrong
  responses. This document defines the approach, test matrix, and operational requirements for closing that gap.

  Key decisions:
  - Tests run on every merge to main via CI (not on every push, not just on demand)
  - Many tests across 8 failure domains: fault tolerance, streaming edge cases, thread/context management, response format, structured outputs,
  tool calls, and LLM-as-judge semantic checks
  - Provider credentials are injected via environment variables; tests skip automatically when absent
  - No Tutor or running LMS required — OpenEdXProcessor is mocked, only the LiteLLM network call is live
  - LLM-as-judge tests use a reasoning-focused model for accuracy, not a speed model

  What this PR does NOT include

  No implementation code — only the ADR. Implementation will follow once the approach is agreed.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @Henrrypg!


This repository is currently maintained by `@felipemontoya`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+The existing test suite exercises every layer of the plugin — orchestrators, processors, API endpoints, session management — but every test mocks the LLM call itself. This means the tests verify that the plugin *plumbs* a request correctly to LiteLLM, but they cannot detect:
+
+- A provider's API rejecting a request because of a changed model name, deprecated parameter, or schema enforcement rule.
+- A structured-output contract (`response_format` / JSON schema) that the LLM silently ignores or returns in an unexpected shape.
+- Thread-based context not surviving across turns on the provider's side.
+- Responses that are syntactically correct but semantically unrelated to the question asked.
+- Silent failures in error-recovery paths (e.g. stale thread IDs, unguarded `json.loads` calls) that only trigger under real API conditions.
+
+These failure modes only surface against a real network call. A lightweight integration suite that runs on demand (not in every CI push) closes this gap without bloating the regular test runtime.
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: We could certainly do better than just on demand. But I agree that every push may be too much. What if we run it every time something is merged to master? We still have to open a new PR to fix, but we would know before doing a release

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    ├── test_educator_assistant.py     # Z, AA  structured-output quality
+    ├── test_tool_calls.py             # AD, AE  tool-call round-trips
+    └── test_semantic_quality.py       # AF, AG, AH  LLM-as-judge extensions
+```
+
+**`integration_test_settings.py`** — A Django settings module that inherits `test_settings` and overrides `AI_EXTENSIONS` with provider entries backed by environment variables (`OPENAI_API_KEY`, `ANTHROPIC_API_KEY`). Only activated when running the integration suite; `make test` continues to use `test_settings`.
+
+**`tests/integration/conftest.py`** — Declares `PROVIDERS` (the parametrised list of provider slugs and their required env-var names), `skip_if_no_key` (runtime skip helper), `create_profile_and_scope` (creates `AIWorkflowProfile` + `AIWorkflowScope` with provider slug injected via `content_patch`), and shared fixtures (`course_key`, `location_id`, `live_user`, `live_api_client`).
+
+The `integration_test_settings.py` targets `openai/gpt-4o-mini` and `anthropic/claude-haiku-3-5` — cheap, fast models appropriate for integration checks.
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: It should target whatever we set as the provider defaults (which by the way we need to update maybe on a montly o bimonthly cadence)

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+### Key design choices
+
+**Automatic skip when keys are absent.** Parametrised tests call `skip_if_no_key` as their first statement; provider-specific tests use `@pytest.mark.skipif` evaluated at collection time. Both produce a clean skip rather than a failure in keyless environments.
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: This is good, but part of this overall issue would be to make sure we administratively have valid keys with tokens for this.

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+**Provider override via `content_patch`.** Rather than creating new profile JSON templates on disk, each test injects the desired provider slug and any extra options into an existing base profile using the RFC 7386 merge-patch mechanism already supported by `AIWorkflowProfile`. This keeps test infrastructure thin and exercises the same config-merge path used in production.
+
+**Bad credentials injected the same way.** Fault-tolerance tests override `options.api_key` or `options.model` via the same `content_patch` / `extra_llm_patch` mechanism, confirming that invalid configuration is surfaced as an error rather than silently producing a `completed` response.
+
+**OpenEdX content is mocked; LLM call is real.** `OpenEdXProcessor` is patched to return a fixed string so tests do not depend on a running LMS. Only the LiteLLM network call is live, isolating the variable under test to the provider response.
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: and importantly, without having to add all the steps to get a running llm with tutor.

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+**Thread and context tests bypass the API layer.** Tests that validate multi-turn behaviour instantiate `LLMProcessor` directly with a `MagicMock` session object. This avoids entanglement with `SubmissionProcessor`, which requires the submissions Django app. The mock session records attribute assignments (`remote_response_id`) so assertions can be made on them directly.
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: SubmissionProcessor is a library, we could install it for CI

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+**`EducatorAssistantProcessor` called directly.** Educator tests instantiate `EducatorAssistantProcessor` with the schema loaded from `response_schemas/educator_quiz_questions.json` passed as `extra_params["response_format"]`, mirroring the production call path in `EducatorAssistantOrchestrator`. This avoids session and submission-processor complexity while exercising the real unguarded `json.loads` at line 83.
+
+**`json.loads` interception.** Test AA patches `json.loads` inside the processor module to capture the raw LLM string before the processor consumes it. The captured string is independently validated, so the test fails with a clear message if the LLM ever returns non-JSON — a failure mode that would otherwise crash the processor with an unhandled `JSONDecodeError`.
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: I dont see the need for this, we control the profiles.json, why should the integration testing avoid loading them as they are?

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+**LLM-as-judge for semantic validation.** Tests F, AF, AG, and AH use a second LLM call as an evaluator: the primary response is fed back to the same provider with a strict system prompt that returns a JSON verdict (`yes` / `no`). This detects responses that are structurally valid but semantically incorrect.
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: LLM with a better model (one focused on train of thought and not speed. E,g, claude 4.8)

**Review by felipemontoya — CHANGES_REQUESTED** [review]

File: docs/decisions/0011-live-llm-provider-integration-tests.md (hunk 1 above)
Comment (Henrrypg) [reply docs/decisions/0011-live-llm-provider-integration-tests.md]: I like this idea, i will update the adr

File: docs/decisions/0011-live-llm-provider-integration-tests.md (hunk 2 above)
Comment (Henrrypg) [reply docs/decisions/0011-live-llm-provider-integration-tests.md]: Got it

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+| # | Test name | Providers | What is validated |
+|---|-----------|-----------|-------------------|
+| A | `test_provider_returns_non_empty_response` | OpenAI, Anthropic | Status is `completed`; response string has more than 10 characters |
+| B | `test_streaming_yields_content` | OpenAI, Anthropic | Response content type is `text/plain`; accumulated streaming bytes exceed 20 characters |
+| C | `test_response_format_json_schema` | OpenAI, Anthropic | Response parses as valid JSON; required key `answer` is present and is a string |
+| D | `test_threaded_stores_remote_response_id` | OpenAI only | `session.remote_response_id` is non-empty after the first `chat_with_context` call |
+| E | `test_threaded_context_maintained_openai` | OpenAI only | A fact planted in turn one is recalled correctly in turn two of the same server-side thread |
+| F | `test_llm_judge_response_relevance` | OpenAI, Anthropic | LLM-as-judge returns `yes` confirming the response is topically consistent with the source content |
+
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md:80]: eventually we might even want a test that looks at the cache usage for claude

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+### Test matrix
+
+#### Baseline — happy path (`test_live_llm_providers.py`)
+
+| # | Test name | Providers | What is validated |
+|---|-----------|-----------|-------------------|
+| A | `test_provider_returns_non_empty_response` | OpenAI, Anthropic | Status is `completed`; response string has more than 10 characters |
+| B | `test_streaming_yields_content` | OpenAI, Anthropic | Response content type is `text/plain`; accumulated streaming bytes exceed 20 characters |
+| C | `test_response_format_json_schema` | OpenAI, Anthropic | Response parses as valid JSON; required key `answer` is present and is a string |
+| D | `test_threaded_stores_remote_response_id` | OpenAI only | `session.remote_response_id` is non-empty after the first `chat_with_context` call |
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: aparently gemini also has a `previous_interaction_id`

Comment (felipemontoya) [reply docs/decisions/0011-live-llm-provider-integration-tests.md]: By default, the API stores all Interaction objects (store=true). On the Paid Tier, the system retains interactions for 55 days; on the Free Tier, for 1 day. If you don't want this, you can set store=false in your request. [Google AI](https://ai.google.dev/gemini-api/docs/interactions/interactions-overview)

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+| H | `test_wrong_model_name_does_not_return_completed` | OpenAI, Anthropic | Error surfaced; no `completed` response with placeholder text | `adapt_to_provider` passes model name through without validation |
+
+#### Streaming edge cases (`test_streaming_edge_cases.py`)
+
+| # | Test name | Providers | What is validated | Risk caught |
+|---|-----------|-----------|-------------------|-------------|
+| K | `test_streaming_handles_empty_delta_chunks` | OpenAI, Anthropic | No crash; no error marker in accumulated bytes | `content = chunk.choices[0].delta.content or ""` yields empty bytes silently |
+| L | `test_streaming_long_response_arrives_completely` | OpenAI, Anthropic | All chunks arrive; final text >200 chars; no error marker | No check that stream fully drains before closing |
+| M | `test_streaming_with_response_format_openai` / `test_streaming_with_response_format_anthropic_clean_outcome` | OpenAI / Anthropic | Valid streamable output OR clean rejection; no 500 crash | OpenAI does not support streaming + strict JSON schema in all model variants; Anthropic rejects it outright |
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: this is something we could map in the _PROVIDER_CAPABILITIES constant and make sure we dont even allow this to be called.

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+| AI | `test_healthy_stream_has_no_error_marker` | OpenAI, Anthropic | `"error_in_stream"` substring NOT in accumulated bytes | `\|\|{...}\|\|` error marker injected on streaming failure; must not appear in healthy call |
+
+#### Thread / context management (`test_threading.py`)
+
+| # | Test name | Providers | What is validated | Risk caught |
+|---|-----------|-----------|-------------------|-------------|
+| N | `test_stale_thread_id_triggers_recovery` | OpenAI only | `previous_response_not_found` caught; stale ID replaced; valid response returned | Recovery hole in `threaded_orchestrator.py:153-159`; stale `remote_response_id` not auto-cleared |
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: given the gemini limit of 55 days, we can search the limit for openai and respond gracefully when expired

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+|---|-----------|-----------|-------------------|-------------|
+| S | `test_response_format_no_extra_keys` | OpenAI, Anthropic | Parsed response has EXACTLY the declared keys; no extras | Schema declared but enforcement depends on model compliance |
+| T | `test_response_format_required_array_non_empty` | OpenAI, Anthropic | Returned array has ≥ 1 element | `educator_quiz_questions.json` has no `minItems`; LLM can return `[]` silently |
+| V | `test_anthropic_streaming_with_strict_schema_no_crash` | Anthropic only | Clean error or graceful fallback; not a 500 crash | Anthropic does not support structured output + streaming in all API versions |
+
+#### Token usage tracking (`test_token_usage.py`)
+
+| # | Test name | Providers | What is validated | Risk caught |
+|---|-----------|-----------|-------------------|-------------|
+| W | `test_usage_populated_after_non_streaming_call` | OpenAI, Anthropic | `usage` dict present; `total_tokens > 0`; `prompt_tokens` and `completion_tokens` both set | `self.usage = None` initial; only set if response carries usage; never validated downstream |
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: I dont know if we should drag this into the code. We already make no promises about total_token count. We could leave that only with unit test_coverage

File: docs/decisions/0011-live-llm-provider-integration-tests.md (hunk 6 above)
Comment (Henrrypg) [reply docs/decisions/0011-live-llm-provider-integration-tests.md]: You're rigth. This adds extra complexity without any value

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+| AG | `test_response_does_not_hallucinate_beyond_content` | OpenAI, Anthropic | "Does the response contain only information from the provided content?" → yes | No grounding check; models confabulate on fictional/narrow content |
+| AH | `test_response_not_truncated_mid_list` | OpenAI, Anthropic | "Does the response mention all five steps?" → yes | `MAX_TOKENS=500` cap may cut responses mid-sentence |
+
+## Consequences
+
+- Any developer with valid API keys can run `make test-integration` from the `backend/` directory and get immediate feedback on provider compatibility.
+- CI pipelines that do not have API keys continue to run `make test` without disruption; live tests are automatically skipped.
+- Adding a new LLM provider to the integration suite requires only a new entry in `PROVIDERS` (in `conftest.py`) and a matching key in `integration_test_settings.py`. No new test functions are needed for the parametrised cases; only thread-specific tests (D, E, N, O, P) may need provider-specific variants.
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md:161]: This is subject to changing a bit once we connect it to github ci

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+- Live tests consume real API credits on every run. Test models are chosen to minimise cost; the `MAX_TOKENS` cap is set conservatively at 500 tokens.
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: except llm-as-judge, this is optimized for quality

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+- The LLM-as-judge pattern (tests F, AF, AG, AH) is probabilistic: a judge model could theoretically return `no` for a correct response. The judge prompt and content are designed to minimise ambiguity, but occasional flakiness on semantically borderline responses should be expected and investigated case by case rather than treated as a systematic failure.
+- Fault-tolerance tests (G, H) assert on error surfacing rather than on a specific HTTP status code, because the exception propagation path depends on the DRF view's error handler. This makes the assertions robust to future view refactors.
+- The Anthropic cache test (Q) requires a large enough context to exceed the provider's 1024-token minimum for cache activation. If the prompt is too short, the test will report `cache_read_input_tokens = 0` without failing, indicating the cache threshold was not met rather than a bug in the integration.
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: worth taking a second look when implmenting and if it is impractical we'll leave it out.

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+
+### Test matrix
+
+#### Baseline — happy path (`test_live_llm_providers.py`)
+
+| # | Test name | Providers | What is validated |
+|---|-----------|-----------|-------------------|
+| A | `test_provider_returns_non_empty_response` | OpenAI, Anthropic | Status is `completed`; response string has more than 10 characters |
+| C | `test_response_format_json_schema` | OpenAI, Anthropic | Response parses as valid JSON; required key `answer` is present and is a string |
+| D | `test_threaded_stores_remote_response_id` | OpenAI only | `session.remote_response_id` is non-empty after the first `chat_with_context` call (uses `MagicMock` session) |
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md]: Will include gemini when we have a key (you could say. providers that support "server_side_thread_id".

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+|---|-----------|-----------|-------------------|-------------|
+| AD | `test_tool_call_pipeline_completes` | OpenAI only | Call completes without hang; non-empty response; usage populated | Tool executor logs errors and continues; no test that tool result actually reaches LLM |
+| AE | `test_unknown_tool_name_returns_error_string` / `test_empty_available_tools_does_not_crash` | OpenAI only | Error string returned; processor completes; no unhandled exception | `tool_executor.py` uses `continue` on unknown tool; silently dropped |
+
+#### Semantic quality (`test_semantic_quality.py`)
+
+| # | Test name | Providers | Judge question | Risk caught |
+|---|-----------|-----------|----------------|-------------|
+| AF | `test_response_language_matches_content` | OpenAI, Anthropic | "Is the response in the same language as the content?" → yes | No language enforcement; model may respond in English for non-English content |
+| AG | `test_response_does_not_hallucinate_beyond_content` | OpenAI, Anthropic | "Does the response contain only information from the provided content?" → yes | No grounding check; models confabulate on fictional/narrow content |
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md:146]: This will most likely be failed by several providers as is. In that case, we might want to mark them optional and continue testing without failing the suite until we can figure out a reasonable solution

File: docs/decisions/0011-live-llm-provider-integration-tests.md [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+| AH | `test_response_not_truncated_mid_list` | OpenAI, Anthropic | "Does the response mention all five steps?" → yes | `MAX_TOKENS=500` cap may cut responses mid-sentence |
```
Comment (felipemontoya) [inline docs/decisions/0011-live-llm-provider-integration-tests.md:147]: Not clear, which five steps. Perhaps there is a bit of explatnation missing

**Review by felipemontoya — APPROVED** [review]
I left some comments that came to mind while reading. Overall I think the ADR is in a good standing and should be merged. We should do an update after the implementation phase to adjust the notes.
